### DISCUSSION TITLE
time to Start saving in transient simulation

### DISCUSSION THREAD

**Discussion by Angelo50** [SpiceSharp/SpiceSharp#181] [General]
Hi again,

Sorry to ask already a new question (I'm doing my best to not ask you too much...be sure ^^

Anyway,  I'm building a front end simulator with beautiful animations
For that,  I launch a transient analysis, save the result in an array and then, run all the simulation objects according to this array.
It works pretty fine, but, running the transient with 0.0001 max step value takes between 5 and ten seconds compute (but it seems I need this value to get correct accuracy) 
So I would Iike to launch the analysis for  the first second result,  save resut to my array, do staff and then relaunch the analysis for the 2nd second (I hope I'm clear..English is not my first language)

Is that feasible?

Thanks

Angelo.

**Comment by svenboulanger** [comment]
Rather than restarting a simulation, you can simply hijack the `ExportSimulationData` event and run the logic there. The simulation pauses execution until all events have finished running.

Kind regards,
Sven

**Reply by Angelo50** [reply]
Sven,

In fact, in a standard Visual application, I'm able to update things inside the simulationExport event loop
In the following code, for each Simulation export result, it launch the function "analyse" which draw two lines linked on the current of  two led  


    using System;
    using System.Collections.Generic;
    using System.ComponentModel;
    using System.Data;
    using System.Drawing;
    using System.Linq;
    using System.Text;
    using System.Windows.Forms;
    using java.lang.management;
    using SpiceSharp;
    using SpiceSharp.Components;
    using SpiceSharp.Simulations;
    using SpiceSharp.ParameterSets;


    namespace testspice2
    {


        public partial class Form1 : Form
        {

            Graphics table;
            SpiceSharp.Circuit ckt;
            Pen pencil;
            Pen D1;
            Pen D2;
            Brush brush;
            Font font;
            Color rubber;
            PointF start_line;
            PointF end_line;
            int panel_height;
            int panel_width;
            float Y;

            public Form1()
            {
                InitializeComponent();
            }

            private void button1_Click(object sender, EventArgs e)
            {

                #region line header
                //
                // draw X oordinate
                //
                table.DrawLine(pencil, start_line, end_line);
                //
                // draw "x" letter at the end of oordinate
                //
                end_line.X -= 20;// decrement x by 20
                table.DrawString("x", font, brush, end_line);
                //
                // Set start and end of Y oordinate line
                //
                start_line.X = 20;
                start_line.Y = 0;

                end_line.X = start_line.X;
                end_line.Y = panel_width;
                //
                // draw Y oordinate
                //
                table.DrawLine(pencil, start_line, end_line);
                //
                // draw "y" letter at the end of oordinate
                //
                table.DrawString("y", font, brush, start_line);
                //
                // arrows are up to you to draw it
                //
                #endregion

                var Ledmodel = new DiodeModel("LedSimulation");
                Ledmodel.SetParameter("is", 93.2e-11);
                Ledmodel.SetParameter("rs", 0.042);
                Ledmodel.SetParameter("n", 3.73);

                var bjtModelNPN = new BipolarJunctionTransistorModel("bjtModel");
                bjtModelNPN.SetParameter("npn", true)

                .SetParameter("is", 2.2e-15)
                .SetParameter("bf", 240.0)
                .SetParameter("cje", 24.9e-12)
                .SetParameter("vje", 1.10)
                .SetParameter("cjc", 12.4e-12)
                .SetParameter("vjc", 0.300)
                .SetParameter("mjc", 0.300)
                .SetParameter("tf", 371e-12)
                .SetParameter("tr", 64e-9)
                .SetParameter("eg", 1.12)
                ;

                // Build the circuit
                ckt = new Circuit(
                       new VoltageSource("V1000", "1", "0", 9),
                       new VoltageSource("V0", "1", "3", 0),
                       new VoltageSource("V1", "3", "11", 0),
                       new VoltageSource("V2", "11", "19", 0),
                       new VoltageSource("V3", "19", "21", 0),
                       new VoltageSource("V4", "21", "13", 0),
                       new VoltageSource("V15", "31", "32", 0),
                       new VoltageSource("V16", "33", "34", 0),
                       new VoltageSource("V17", "35", "36", 0),
                       new VoltageSource("V18", "37", "38", 0),
                       new VoltageSource("V19", "39", "41", 0),
                       new VoltageSource("V20", "41", "45", 0),
                       new VoltageSource("V21", "43", "45", 0),
                       new VoltageSource("V22", "45", "47", 0),
                       new VoltageSource("V23", "47", "0", 0),
                       new Diode("D5", "11", "15", "LedSimulation"),
                       new Diode("D6", "13", "17", "LedSimulation"),
                       new Resistor("R7", "15", "31", 1000),
                       new Resistor("R8", "17", "33", 1000),
                       new Resistor("R9", "19", "37", 220000),
                       new Resistor("R10", "21", "35", 180000),
                       new Capacitor("C11", "37", "31", 0.00002),
                       new Capacitor("C12", "35", "33", 0.00002),
                       new BipolarJunctionTransistor("Q13", "32", "36", "43", "0", "bjtModel"),
                       new BipolarJunctionTransistor("Q14", "34", "38", "39", "0", "bjtModel")
                       );

                ckt.Add(Ledmodel);
                ckt.Add(bjtModelNPN);

                //var dc = new Transient("tran", 1, 100,0.1);
                Y = 228;
                PointF SLine1 = start_line;
                PointF SLine2 = start_line;
                //-------------------------------------------------------------------------------------------------------------------

                var dc = new Transient("tran", 0, 10, 0.0001);
                // Create exports
                var inputExport = new RealVoltageExport(dc, "in").Value;
                Console.WriteLine(inputExport);
                var outputExport = new RealVoltageExport(dc, "out");

                analyse(dc, 0, 0, SLine1, SLine2);


                dc.AfterExecute += (bsender, exportDataEventArgs) =>
                {
                        // MessageBox.Show(cycle.ToString());
                        //Console.WriteLine(cycle);
                    };

                // Run the simulation


                dc.Run(ckt);

            }


            private void button2_Click(object sender, EventArgs e)
            {

            }

            private void Form1_Load(object sender, EventArgs e)
            {
                //
                // Set new graphics that uses panel as drawing surface
                //
                table = panel1.CreateGraphics();
                //
                // Set new PointF(x,y),  start line point 
                // 
                start_line = new PointF();
                start_line.X = 0;
                start_line.Y = 0;
                //
                // Set new PointF(x,y) end line point
                // 
                end_line = new PointF();
                end_line.X = 0;
                end_line.Y = 0;
                //
                // Set new Pen, 
                // pencil that will be used for 
                // drawing with Red color and 2 pixels tick
                //
                pencil = new Pen(Color.Black, 1);
                D1 = new Pen(Color.Red, 1);
                D2 = new Pen(Color.Blue, 1);
                //
                // Set new Brush
                // brush that will be used for drawing letters
                //
                brush = new SolidBrush(Color.Black);
                //
                // Set new Font
                // font that shall be used for drawing strings
                //
                font = new Font("LucidaConsole", 12, FontStyle.Bold);
                //
                // Set new Color, 
                // rubber that will be used for erasing table
                //
                rubber = new Color();
                rubber = Color.White;
                //
                // Get Panel height and width
                //
                panel_height = panel1.Height;
                panel_width = panel1.Width;
                //
                // Clear table surface with rubber, 
                // rubber contains value of collor 
                // and it must be the same as panel1 background color
                //
                table.Clear(rubber);
                //
                // Set start and end of X oordinate line
                //
                start_line.X = 0;
                start_line.Y = panel_height - 20;

                end_line.X = panel_width;
                end_line.Y = start_line.Y;


            }



            private void analyse(Transient dc, int cycle, int sec, PointF SLine1, PointF SLine2)
            {

                dc.ExportSimulationData += (bsender, exportDataEventArgs) =>
                {
                    PointF point1 = new PointF(SLine1.X + 0.005f, Y - (float)new RealPropertyExport(dc, "D5", "i").Value * 15000);
                    PointF point2 = new PointF(SLine2.X + 0.005f, Y - (float)new RealPropertyExport(dc, "D6", "i").Value * 15000);
                    Console.WriteLine(new RealPropertyExport(dc, "D5", "i").Value);
                    table.DrawLine(D1, SLine1, point1);
                    table.DrawLine(D2, SLine2, point2);
                    SLine1 = point1;
                    SLine2 = point2;
                    cycle++;
                };
            }
        }
    }


But I work in a Unity project and the loop inside the Simulation Export Event is done in the same frame.
This means that All unity updates are not executed until the end of the simulation export.

This is why I decided to save the simulation export inside an array and the , after, update things at the time I want according this array
but I would prefer being able to update things directy in the simulation export event

If you or Someone else could help

Thanks again for you time

Angelo.

**Reply by marcin-golebiowski** [reply]
I would suggest you to :

- use additional thread for SpiceSharp simulation (and save the results in ExportSimulationData  to some thread-safe structure)  
- render the data from that structure in Unity

**Reply by Angelo50** [reply]
Thanks for your answer Marcin

Could you please tell me more about or may be provide some piece of code to enlight me ?

Thanks again
Angelo.

**Reply by svenboulanger** [reply]
I agree with Marcin, you will probably want to run the simulation in a separate thread that will be independent from your Unity game loop. Perhaps you can start your quest [here](https://www.tutorialspoint.com/csharp/csharp_multithreading.htm) and [here](https://www.c-sharpcorner.com/UploadFile/de41d6/monitor-and-lock-in-C-Sharp/).

**Reply by Angelo50** [reply]
Hi,
First, thanks for your time and help.
I finally solved my problem with you help.
My solution is the following:
I launch the Transient analysis in a separate thread, then for each simulation Event, I copy the result in a matrix, (I didn't find a better way to keep the simulation result...but it works) 
Then, I pause the thread, initiate the unity rendering and when's done,  resume the Event thread for the next result.
This allow me to get a kind of interactive solution to render at each simulation event

Best Regards

Angelo.
